### PR TITLE
support JVM attack

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
support JVM attack by [byteman](https://github.com/bytemanproject/byteman), and there is a simple demo about how to use byteman [byteman-example](https://github.com/WangXiangUSTC/byteman-example)

byteman don't support inject stress on cpu or memory, implement it in [#1](https://github.com/WangXiangUSTC/byteman/pull/1)

## Usage

```bash
./bin/chaosd attack jvm install --help
install agent to Java process

Usage:
  chaosd attack jvm install [options] [flags]

Flags:
  -h, --help       help for install
      --pid int    the pid of Java process which need to attach
      --port int   the port of agent server (default 9288)

Global Flags:
      --log-level string   the log level of chaosd, the value can be 'debug', 'info', 'warn' and 'error'
```

```bash
./bin/chaosd attack jvm submit --help 
submit rules for byteman agent

Usage:
  chaosd attack jvm submit [command]

Available Commands:
  exception   throw specified exception for specified method
  gc          trigger GC for JVM
  latency     inject latency to specified method
  return      return specified value for specified method
  stress      inject stress to JVM

Flags:
  -h, --help          help for submit
  -n, --name string   rule name, should be unique, and will generate one automatically if it is empty
      --port int      the port of agent server (default 9288)

Global Flags:
      --log-level string   the log level of chaosd, the value can be 'debug', 'info', 'warn' and 'error'

Use "chaosd attack jvm submit [command] --help" for more information about a command.
```

## Examples:

Run java application in [byteman-example](https://github.com/WangXiangUSTC/byteman-example):

```bash
 java -jar ./example.helloworld/Main.jar
```

You can see the output:

```log
0. Hello World
1. Hello World
2. Hello World
3. Hello World
...
```

Get the PID of this process, it is 20350 when I'm running.

Then attach agent to this process:

```bash
./bin/chaosd attack jvm install --pid 20350 --port 9288
```

Inject exception for the function `Main.sayhello`:
```bash
./bin/chaosd attack jvm submit exception -c Main -m sayhello --name sayhello-exception --port 9288 --exception 'java.io.IOException("BOOM")'
```

You can see exception in the log:

```log
Rule.execute called for sayhello-exception_0:0
sayhello-exception execute
caught ThrowException
Got an exception!java.io.IOException: BOOM
```

Signed-off-by: xiang <xiang13225080@163.com>

### CODE REVIEW

File: pkg/server/chaosd/jvm_attack.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	if err := tmpfile.Close(); err != nil {
+		return err
+	}
+
+	log.Info("create btm file", zap.String("file", tmpfile.Name()))
+
+	bmSubmitCmd := fmt.Sprintf(bmSubmitCommand, attack.Port, "u", tmpfile.Name())
+	cmd := exec.Command("bash", "-c", bmSubmitCmd)
```
Comment (Andrewmatilde) [inline pkg/server/chaosd/jvm_attack.go:219]: I know `sh` is better than `bash` now.[Difference between sh and bash](https://stackoverflow.com/questions/5725296/difference-between-sh-and-bash)

File: pkg/utils/util.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"math/rand"
+	"time"
+)
+
+const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
+
+func RandomStringWithCharset(length int) string {
+	var seededRand = rand.New(rand.NewSource(time.Now().UnixNano()))
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/utils/util.go:24]: @muse-dev help

File: pkg/server/chaosd/jvm.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+package chaosd
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"text/template"
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:23]: @muse-dev ignore

File: pkg/utils/util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/utils/util.go:24]: @muse-dev ignore

File: pkg/server/chaosd/jvm.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	} else {
+		return j.installRule(attack)
+	}
+}
+
+func (j jvmAttack) attach(attack *core.JVMCommand) error {
+	var err error
+
+	bmInstallCmd := fmt.Sprintf(bmInstallCommand, attack.Port, attack.Pid)
+	cmd := exec.Command("bash", "-c", bmInstallCmd)
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:70]: @muse-dev ignore

File: pkg/server/chaosd/jvm_attack.go (hunk 1 above)
Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm_attack.go]: but in chaos mesh and chaosd, we all use bash now, maybe we can change them all later

**Review by Andrewmatilde — COMMENTED** [review]
I think unit test is nesscesary on jvm attack.

File: cmd/attack/jvm.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+			options.Type = core.JVMInstallRuleType
+			options.CompleteDefaults()
+			utils.FxNewAppWithoutLog(dep, fx.Invoke(jvmCommandFunc)).Run()
+		},
+	}
+
+	cmd.Flags().StringVarP(&options.Name, "name", "n", "", "rule name, should be unique, and will generate one automatically if it is empty")
+	cmd.Flags().StringVarP(&options.Class, "class", "c", "", "Java class name")
+	cmd.Flags().StringVarP(&options.Method, "method", "m", "", "the method name in Java class")
+	cmd.Flags().StringVarP(&options.Action, "action", "a", "", "fault action, values can be latency, exception, return or stress")
```
Comment (Andrewmatilde) [inline cmd/attack/jvm.go:81]: I suggest to splite different action in to different command.
And 'gc' should add into description.

File: cmd/attack/jvm.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	cmd.Flags().IntVarP(&options.Port, "port", "", 9288, "the port of agent server")
```
Comment (Andrewmatilde) [inline cmd/attack/jvm.go:82]: Uint16 may be better , but not necessary.

File: pkg/server/chaosd/jvm.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+		return j.installRule(attack)
+	}
+
+	return errors.Errorf("attack type %s not supported", attack.Type)
+}
+
+func (j jvmAttack) attach(attack *core.JVMCommand) error {
+	var err error
+
+	bmInstallCmd := fmt.Sprintf(bmInstallCommand, attack.Port, attack.Pid)
```
Comment (Andrewmatilde) [inline pkg/server/chaosd/jvm.go:78]: func attach here use install bash command , but func install rule use submit bash command. I think the naming here is not proper.

File: pkg/server/chaosd/jvm.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+RULE {{.Name}}
+GC
+ENDRULE
+`
+
+type jvmAttack struct{}
+
+var JVMAttack AttackType = jvmAttack{}
+
+const bmInstallCommand = "bminstall.sh -b -Dorg.jboss.byteman.transform.all -Dorg.jboss.byteman.verbose -p %d %d"
```
Comment (Andrewmatilde) [inline pkg/server/chaosd/jvm.go:60]: I suggest to give way to install byteman in chaosd , or auto install it or support use path of file such as 'bminstall.sh'.Anyway,I think code in chaosd dependent on configure in other repo is not a good idea.
`bminstall.sh` here rely on ```export BYTEMAN_HOME=`pwd`/byteman-download-4.0.14
export PATH=$PATH:${BYTEMAN_HOME}/bin``` in byteman setting.

File: cmd/attack/jvm.go (hunk 5 above)
Comment (WangXiangUSTC) [reply cmd/attack/jvm.go]: Good suggestion, done in fba7031

File: pkg/server/chaosd/jvm.go (hunk 7 above)
Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:78]: rename it in https://github.com/chaos-mesh/chaosd/pull/33/commits/49f24410a36dcbfbc842ace7b2b52ba5e530e2ad

Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:78]: rename it in https://github.com/chaos-mesh/chaosd/pull/33/commits/49f24410a36dcbfbc842ace7b2b52ba5e530e2ad

Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:78]: rename it in https://github.com/chaos-mesh/chaosd/pull/33/commits/49f24410a36dcbfbc842ace7b2b52ba5e530e2ad

File: pkg/server/chaosd/jvm.go (hunk 8 above)
Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:60]: I think we can involve all the tools or libs required for chaosd in it's download package file. The package may looks like:

chaosd-linux-v0.9.0
     |── bin
     │   └── chaosd
     |── tools
         |── byteman
         |── stress-ng

Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:60]: @cwen0 Do you think it's OK?

Comment (cwen0) [reply pkg/server/chaosd/jvm.go:60]: Can we move out `chaosd` binary from `bin` directory?

Comment (WangXiangUSTC) [reply pkg/server/chaosd/jvm.go:60]: It look good to me

**Comment by WangXiangUSTC** [comment]
@Andrewmatilde @Gallardot PTAL

**Review by Gallardot — COMMENTED** [review]
Rest LGTM

File: test/integration_test/jvm/run.sh [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+$bin_path/chaosd attack jvm submit return --class Main --method getnum --port 9288  --value 99999
+check_contains "99999" helloworld.log
+
+$bin_path/chaosd attack jvm submit exception  --class Main --method sayhello --port 9288 --exception 'java.io.IOException("BOOM")'
+check_contains "BOOM" helloworld.log
+
+# TODO: add test for latency, stress and gc
+
+echo "clean"
+kill $pid
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply test/integration_test/jvm/run.sh]: done

File: test/utilities/check_contains [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+set -eu
+
+if ! grep -Fq "$1" "$2"; then
+    echo "TEST FAILED: $2 DOES NOT CONTAIN '$1'"
+    echo "____________________________________"
+    cat $2
+    echo "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^"
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply test/utilities/check_contains]: done

File: pkg/core/jvm.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+			}
+
+			if len(j.Method) == 0 {
+				return errors.New("method not provided")
+			}
+
+		}
+
+		if len(j.Name) == 0 {
+			j.Name = fmt.Sprintf("%s-%s-%s-%s-%s", j.Class, j.Method, j.Action, j.Type, utils.RandomStringWithCharset(5))
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/core/jvm.go:126]: I remove it from the flag, and still keep it here

File: pkg/core/jvm.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	if j.Type == JVMInstallType {
+		if j.Pid == 0 {
+			return errors.New("pid can't be 0")
+		}
+
+	} else if j.Type == JVMSubmitType {
+		if len(j.Action) == 0 {
+			return errors.New("action not provided, action can be 'latency', 'exception', 'return', 'stress' or 'gc'")
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/core/jvm.go:97]: done

**Comment by WangXiangUSTC** [comment]
@Gallardot PTAL again

**Review by Gallardot — APPROVED** [review]
LGTM

**Comment by WangXiangUSTC** [comment]
@Gallardot Thanks!

File: cmd/attack/jvm.go [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+	return cmd
+}
+
+func NewJVMInstallCommand(dep fx.Option, options *core.JVMCommand) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "install [options]",
+		Short: "install agent to Java process",
+		Run: func(*cobra.Command, []string) {
+			options.Type = core.JVMInstallType
+			options.CompleteDefaults()
```
Comment (Andrewmatilde) [inline cmd/attack/jvm.go:56]: It seems that core.JVMCommand not implement this func. Is it really necessary here ?

**Review by Andrewmatilde — COMMENTED** [review]
Rest LGTM

File: cmd/attack/jvm.go (hunk 13 above)
Comment (WangXiangUSTC) [reply cmd/attack/jvm.go]: I think it's better to remain it, and I implement this function in 3a80f1e

Comment (Andrewmatilde) [reply cmd/attack/jvm.go]: OK

**Pull request merged**